fs/virtualfs/virtualfs.go

```diff
@@ -94,6 +99,73 @@ func NewStaticDirectory(name string, entries []fs.Entry) fs.Directory {
 	}
 }
 
+type streamingDirectory struct {
+	virtualEntry
+	// Used to generate the next entry and execute the callback on it.
+	// +checklocks:mu
+	callback           func(context.Context, func(context.Context, fs.Entry) error) error
+	multipleIterations bool
+	mu                 sync.Mutex
+}
+
+var errChildNotSupported = errors.New("streamingDirectory.Child not supported")
+
+func (sd *streamingDirectory) Child(ctx context.Context, name string) (fs.Entry, error) {
+	return nil, errChildNotSupported
+}
+
+var errIteratorAlreadyUsed = errors.New("cannot use streaming directory iterator more than once if not MultipleIterations") // +checklocksignore: mu
+
+func (sd *streamingDirectory) getIterator() (func(context.Context, func(context.Context, fs.Entry) error) error, error) {
+	sd.mu.Lock()
+	defer sd.mu.Unlock()
+
+	if sd.callback == nil {
+		return nil, errIteratorAlreadyUsed
+	}
+
+	cb := sd.callback
+
+	if !sd.MultipleIterations() {
+		sd.callback = nil
```

Comment:
why provide this flexibility i.e single or multiple pass of reading entries of a directory ? Since it is a Stream directory by not make it single pass read ? Just to avoid complexity. Even if we provide this flexibility here, can we provide the same for stream files and how can we enforce that multiple pass directory have only multiple pass file and visa versa